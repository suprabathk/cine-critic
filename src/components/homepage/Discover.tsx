import { getDiscoverList } from "@/utils/movie-utils";
import { useQuery } from "@tanstack/react-query";
import MovieCard from "@/components/homepage/MovieCard";
import { MoviePreview } from "@/types/movies";
import Grid from "@mui/material/Grid";
import { Typography } from "@mui/material";
import { MoviesSkeleton } from "@/loadingSkeletons/homepage";

const Discover = () => {
  const discoverQuery = useQuery({
    queryFn: getDiscoverList,
    queryKey: ["discover"],
  });

  return (
    <div style={{ width: "100%", marginTop: "2rem" }}>
      <Typography
        variant="h4"
        component="h4"
        fontWeight={700}
        lineHeight="2.25rem"
        fontSize="1.5rem;"
        color="inherit"
        marginBottom="0.5rem"
      >
        Discover
      </Typography>
      {discoverQuery.isLoading ? (
        <MoviesSkeleton />
      ) : (
        <Grid container justifyContent="space-between" gap={2}>
          {discoverQuery.data.results.map((movie: MoviePreview) => (
            <Grid key={movie.id}>
              <MovieCard movie={movie} />
            </Grid>
          ))}
        </Grid>
      )}
    </div>
  );
};

export default Discover;
