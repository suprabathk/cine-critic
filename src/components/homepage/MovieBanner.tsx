import * as React from "react";
import { MoviePreview } from "@/types/movies";
import { useRouter } from "next/router";
import movieStyles from "@/styles/movie.module.css";
import { CalendarMonthOutlined, People } from "@mui/icons-material";
import { Stack, Typography, Chip, Rating } from "@mui/material";

export default function MovieBanner(props: { movie: MoviePreview }) {
  const router = useRouter();
  const {
    id,
    title,
    overview,
    poster_path,
    backdrop_path,
    adult,
    release_date,
    vote_average,
    vote_count,
  } = props.movie;

  return (
    <div
      className={movieStyles.movieBanner}
      onClick={() => router.push(`/movie/${id}`)}
      style={{
        backgroundImage: `linear-gradient(rgba(0, 0, 0, 0.9), rgba(0, 0, 0, 0.9)), url(${process.env.NEXT_PUBLIC_TMDB_API_IMAGES}/${backdrop_path})`,
      }}
    >
      <Stack
        direction="row"
        alignItems="center"
        justifyContent="center"
        width="100%"
      >
        <img
          src={`${process.env.NEXT_PUBLIC_TMDB_API_IMAGES}/${poster_path}`}
          alt="poster"
          className={movieStyles.bannerPoster}
        />
        <Stack height="100%" alignItems="start" marginLeft="2rem">
          <Typography
            variant="h4"
            fontWeight="600"
            textAlign="center"
            marginTop="1rem"
            lineHeight="1.5rem"
          >
            {title}
          </Typography>
          <Typography variant="body2" marginTop="0.6rem">
            {overview}
          </Typography>
          <Stack
            direction="row"
            alignItems="center"
            width="100%"
            justifyContent="space-between"
            marginTop="1rem"
          >
            <Chip
              label={release_date}
              icon={<CalendarMonthOutlined />}
              size="small"
            />
            <Stack direction="row" alignItems="center" gap="0.5rem">
              <Rating size="small" value={vote_average! / 2} readOnly />
              <Chip icon={<People />} label={vote_count} size="small" />
            </Stack>
          </Stack>
        </Stack>
      </Stack>
    </div>
  );
}
