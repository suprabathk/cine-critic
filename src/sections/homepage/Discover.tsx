import { MoviePreview } from "@/types/movies";
import Grid from "@mui/material/Grid";
import { Typography } from "@mui/material";
import { MoviesSkeleton } from "@/loadingSkeletons/homepage";
import MovieCard from "@/components/homepage/MovieCard";
import useDiscoverQuery from "@/hooks/useDiscoverQuery";

const Discover = () => {
  const discoverQuery = useDiscoverQuery();

  return (
    <div style={{ width: "100%", marginTop: "2rem" }}>
      <Typography
        variant="h4"
        component="h4"
        fontWeight={700}
        lineHeight="2.25rem"
        fontSize="1.5rem;"
        color="inherit"
        marginBottom="0.5rem"
      >
        Discover
      </Typography>
      {discoverQuery.isLoading ? (
        <MoviesSkeleton />
      ) : (
        <Grid container spacing="1rem">
          {discoverQuery.data.results.map((movie: MoviePreview) => (
            <Grid item key={movie.id} xs={12} md={6} lg={4}>
              <MovieCard movie={movie} />
            </Grid>
          ))}
        </Grid>
      )}
    </div>
  );
};

export default Discover;
