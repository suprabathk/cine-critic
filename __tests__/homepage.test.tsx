import MovieBanner from "@/components/homepage/MovieBanner";
import ActionAreaCard from "@/components/homepage/MovieCard";
import Discover from "@/sections/homepage/Discover";
import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

afterEach(() => {
  jest.clearAllMocks();
});

const dummyMovieData = {
  adult: false,
  backdrop_path: "",
  genre_ids: [],
  id: 0,
  original_language: "",
  original_title: "",
  overview: "",
  popularity: 0,
  poster_path: "",
  release_date: "",
  title: "movieTitle",
  vote_average: 0,
  vote_count: 0,
};

jest.mock("../src/utils/auth-utils", () => ({
  isAuth: jest.fn().mockReturnValue(true),
}));

const pushMock = jest.fn();
jest.mock("next/router", () => ({
  useRouter() {
    return {
      push: pushMock,
    };
  },
}));

jest.mock("../src/hooks/useDiscoverQuery", () => ({
  useDiscoverQuery: jest.fn().mockReturnValue({
    data: {
      results: [
        {
          adult: false,
          backdrop_path: "string",
          genre_ids: [],
          id: 1,
          original_language: "original_language",
          original_title: "original_title",
          overview: "overview",
          popularity: 0,
          poster_path: "",
          release_date: "release_date",
          title: "title",
          vote_average: 0,
          vote_count: 0,
        },
      ],
    },
    isLoading: false,
    error: {},
  }),
}));

describe("Homepage tests", () => {
  it("Snapshot testing", async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(<Discover />);
    });
    expect(tree!.toJSON()).toMatchSnapshot();
  });

  describe("Movie banner tests", () => {
    it("Redirect on click", async () => {
      render(<MovieBanner movie={dummyMovieData} />);
      const titleText = screen.getByRole("heading", {
        name: /movieTitle/i,
      });
      await userEvent.click(titleText);
      await waitFor(() => {
        expect(pushMock).toBeCalledTimes(1);
        expect(pushMock).toBeCalledWith("/movie/0");
      });
    });
  });

  describe("Movie card tests", () => {
    it("Redirect on click", async () => {
      const { container } = render(<ActionAreaCard movie={dummyMovieData} />);
      const movieCard = container.getElementsByClassName("movieCardAction")[0];
      await userEvent.click(movieCard);
      await waitFor(() => {
        expect(pushMock).toBeCalledWith("/movie/0");
      });
    });
  });
});
