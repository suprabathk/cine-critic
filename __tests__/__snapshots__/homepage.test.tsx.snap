// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage tests Snapshot testing 1`] = `
<div
  style={
    {
      "marginTop": "2rem",
      "width": "100%",
    }
  }
>
  <h4
    className="MuiTypography-root MuiTypography-h4 css-fxl47-MuiTypography-root"
  >
    Discover
  </h4>
  <div
    className="MuiGrid-root MuiGrid-container css-4585yq-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-lg-4 css-1e6jbaj-MuiGrid-root"
    >
      <div
        className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root movieCard css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <button
          className="MuiButtonBase-root MuiCardActionArea-root movieCardAction css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <img
            alt="Poster"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            height="140"
            src="http://image.tmdb.org/t/p/w500"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1c47xw2-MuiTypography-root"
            >
              title
            </div>
            <p
              className="MuiTypography-root MuiTypography-body2 css-l634b0-MuiTypography-root"
            >
              overview
            </p>
            <div
              className="MuiStack-root css-1tqtcgg-MuiStack-root"
            >
              <div
                className="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-13htn8f-MuiChip-root"
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                style={
                  {
                    "marginTop": "1rem",
                  }
                }
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconSmall MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
                  data-testid="CalendarMonthOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 4h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V10h14v10zm0-12H5V6h14v2zM9 14H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2zm-8 4H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2z"
                  />
                </svg>
                <span
                  className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                >
                  release_date
                </span>
              </div>
              <div
                className="MuiStack-root css-m69qwo-MuiStack-root"
              >
                <span
                  aria-label="0 Stars"
                  className="MuiRating-root MuiRating-sizeMedium Mui-readOnly MuiRating-readOnly css-ryrseu-MuiRating-root"
                  onMouseLeave={[Function]}
                  onMouseMove={[Function]}
                  role="img"
                >
                  <span>
                    <span
                      className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarBorderIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarBorderIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarBorderIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarBorderIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                        data-testid="StarBorderIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
                <div
                  className="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-13htn8f-MuiChip-root"
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                >
                  <span
                    className="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                  >
                    0
                  </span>
                </div>
              </div>
            </div>
          </div>
          <span
            className="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
